### DISCUSSION TITLE
Table enhancements

### DISCUSSION THREAD

**Discussion by ElPiloto** [MunifTanjim/nui.nvim#298] [Ideas]
Hello,

I had a few suggestions for enhancements around the table component.  This came up because I'm trying to use a nui table as a (very limited) spreadsheet where you can edit different cells (see [attached](https://github.com/MunifTanjim/nui.nvim/assets/629190/27e43b94-47a3-4a91-983e-565aa3f1151c)).  I've cobbled something together that works well enough, but it's pretty brittle.  I think a proper implementation could benefit from some of the enhancements listed below.  Thanks for your time and for nui.nvim: it's a real gem for the neovim community!

Here are some "transparency" API enhancements.  These essentially are things that I think are already present in the table component, but could be exposed as part of a public interface in order to make it easier to interact with tables.
* get first non-header row position
* get next cell in row position
* get next row position
* get rendered table width (this would be helpful in knowing how large to make a pop-up containing said table)

A related set of enhancements would be to directly provide navigation functionality for a table:
* move to header
* move to first non-header row
* move to first cell in row
* move to next cell in row
* move to next row
* prevent movement to header sections

Style enhancements:
* allow specifying custom row or cell highlight groups when the cursor is in the current row/cell (not 100% sure how this would work, but it would be great to be able to highlight the current cell)

Tying those requests back to the attached video, the overall functionality would be improved if:
* the pop-up for the table automatically matched the width of the table
* when the table popped up, the cursor started in the first cell location (I know I can manually hard code the start location at row 7, but this seems brittle)
* movement worked between cells instead of letting the cursor land on the table borders

**Comment by ElPiloto** [comment]
Any suggestions on this?

**Reply by MunifTanjim** [reply]
Hey, thanks for the interest. The current table implementation is very basic. I have plans to do most of the things you listed above. I just haven't managed the time to do it yet. But these are coming, that's sure.
